pkg/localstore/mode_put.go

```
@@ -248,20 +239,51 @@ func (db *DB) put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk)
 	return exist, nil
 }
 
-// putSharky will add the item to sharky storage if it doesnt exist.
-func (db *DB) putSharky(ctx context.Context, item shed.Item) (loc sharky.Location, exists bool, err error) {
-	exists, err = db.retrievalDataIndex.Has(item)
+// checkAndRemoveStampIndex will check if we have the postageIndexIndex already taken
+// for a particular {BatchID, BatchIndex}. If yes and the batch is immutable, we
+// return error, if the batch is not immutable we replace the index to point to the
+// new chunk if the timestamp of the new chunk is later.
+// If the index is not taken, we do nothing. This is done to guard against
+// overissuance of batches.
+func (db *DB) checkAndRemoveStampIndex(
+	item shed.Item,
+	batch *leveldb.Batch,
+	loc *releaseLocations,
+) (int64, error) {
+	previous, err := db.postageIndexIndex.Get(item)
+	if errors.Is(err, leveldb.ErrNotFound) {
+		return 0, nil
+	}
+	if err != nil {
+		return 0, fmt.Errorf("failed reading postageIndexIndex: %w", err)
+	}
+	if item.Immutable {
+		return 0, ErrOverwriteImmutable
+	}
+	// if a chunk is found with the same postage stamp index,
+	// replace it with the new one only if timestamp is later
+	if !later(previous, item) {
+		return 0, ErrOverwrite
+	}
+
+	// remove older chunk
+	previousIdx, err := db.retrievalDataIndex.Get(previous)
```

reminder to ignore error here